solo-io/gloo
# Stability: Simplify Image Building and Publishing CI

# Description

Simplify how we setup a kind cluster used in our kubernetes regression tests

# Context
Enterprise equivalent work can be found here: https://github.com/solo-io/solo-projects/pull/4743

## Setup Kind Script
KIND had [a bug](https://github.com/kubernetes-sigs/kind/issues/2402) which caused docker images to fail to load when using "kind load".

This impacted (some) developers ability to run regression tests locally. In these tests, we create a kind cluster, build all of our images, and load them into the cluster, before executing the tests. Our solution was to allow developers to stand up a local image registry, configure kind to pull images from there, and build/tag our images for that registry. 

This was super effective at enabling local development for developers, but it came at the cost of complexity in our CI pipeline. Because we had parallel paths (some who use the local registry and some who use kind-load) it's been challlenging to keep all commands in our Makefile functioning for both cases.

The kind bug is resolved, so this work attempts to simplify our pipeline and ensure developers all use the same code paths.

## Gloomtls Test Suite
This suite contained references to installing gloo using a custom registry. Since we removed the need for that (see above section for description), I simplified these tests.

See https://github.com/solo-io/gloo/pull/7882 for an outline of how we did this previously for other test suites

## Makefile Simplications
We had a number of make targets that did the same thing (build/push docker images).  I simplified this, in an effort to make our makefiles more readable and only contain code that is executed by developers.

The approach was to start with frequently used commands that we could templatize:
- kind-load-{IMAGE}
- docker-push-{IMAGE}

Then I built targets which performed those actions for all images in our project:
- docker (calls each docker-{IMAGE} target)
- docker-push (calls each docker-push-{IMAGE} target
- kind-build-and-load-images (calls docker and each kind-load-{IMAGE} target)

I also renamed the target we use in our release pipeline to: `docker-publsh` and `docker-publish-retag` to clarify when it should be executed and included a comment to developers so they know which targets are intended to be run by CI

## Extended Images
[Internal Slack Ctx](https://solo-io-corp.slack.com/archives/C03MFATU265/p1680799595431999?thread_ts=1680703414.051189&cid=C03MFATU265)

We had removed the publishing of extended images in Enterprise, so I mirrored these changes in open source. Our working assumption is that if we decide to re-add these, it will be easy to do, but in the meantime we are removing unused code.

## Image Repo Rename
https://github.com/solo-io/solo-projects/pull/4743#discussion_r1170183300 spoke with @ianmacclancy and we decide to use image_registry where image_repo had been used. To be consistent we're renaming this across gloo and gloo-enterprise.

#  Testing
All developers should be able to:
- Use the readme to understand how to setup and run kube2e tests
- Execute the setup-kind script and pass any custom values
- Run the regression suites (kube2e) locally
- Build all images (make docker)
- Retag images (make docker-retag-*)

# Test Release
We have some scripts to allow manually triggering a "dev release" (publishing assets) from a branch. I used this workflow (from my branch, and pointing to my gloo open source branch) to initiate a build.

Logs: https://github.com/solo-io/solo-projects/actions/runs/4678480979
Gloo OSS Release Logs: https://github.com/solo-io/solo-projects/actions/runs/4678480979/jobs/8287263387

```
Successfully published a test build of gloo!
....

Can Install Via:
❯ helm repo add gloo-test https://storage.googleapis.com/gloo-ee-test-helm
❯ helm repo update
❯ helm install -n gloo-system gloo-test gloo-test/gloo --create-namespace --version 1.15.0-beta1-bstabilitysimplify-image-registry-c843e4d
```

I followed the commands, and installed Gloo locally.
```
 glooctl check
Checking deployments... OK
Checking pods... OK
Checking upstreams... OK
Checking upstream groups... OK
Checking auth configs... OK
Checking rate limit configs... OK
Checking VirtualHostOptions... OK
Checking RouteOptions... OK
Checking secrets... OK
Checking virtual services... OK
Checking gateways... OK
Checking proxies... OK
No problems detected.
Skipping Gloo Instance check -- Gloo Federation not detected
```




# Checklist:
